packages/grafana-ui/src/components/Sparkline/Sparkline.tsx

```diff
@@ -0,0 +1,166 @@
+import React, { PureComponent } from 'react';
+import {
+  compareDataFrameStructures,
+  DefaultTimeZone,
+  FieldSparkline,
+  IndexVector,
+  DataFrame,
+  FieldType,
+  getFieldColorModeForField,
+  FieldConfig,
+} from '@grafana/data';
+import { AxisPlacement, DrawStyle, GraphFieldConfig, PointVisibility } from '../uPlot/config';
+import { UPlotConfigBuilder } from '../uPlot/config/UPlotConfigBuilder';
+import { UPlotChart } from '../uPlot/Plot';
+import { Themeable } from '../../types';
+
+export interface Props extends Themeable {
+  width: number;
+  height: number;
+  config?: FieldConfig<GraphFieldConfig>;
+  sparkline: FieldSparkline;
+}
+
+interface State {
+  data: DataFrame;
+  configBuilder: UPlotConfigBuilder;
+}
+
+const defaultConfig: GraphFieldConfig = {
+  drawStyle: DrawStyle.Line,
+  showPoints: PointVisibility.Auto,
+  axisPlacement: AxisPlacement.Hidden,
+};
+
+export class Sparkline extends PureComponent<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    const data = this.perepareData(props);
+    this.state = {
+      data,
+      configBuilder: this.prepareConfig(data, props),
+    };
+  }
+
+  componentDidUpdate(oldProps: Props) {
+    if (oldProps.sparkline !== this.props.sparkline) {
+      const data = this.perepareData(this.props);
+      if (!compareDataFrameStructures(this.state.data, data)) {
+        const configBuilder = this.prepareConfig(data, this.props);
+        this.setState({ data, configBuilder });
+      } else {
+        this.setState({ data });
+      }
+    }
+  }
+
+  perepareData = (props: Props): DataFrame => {
+    const { sparkline } = this.props;
+    const length = sparkline.y.values.length;
+    const yFieldConfig = {
+      ...sparkline.y.config,
+      ...this.props.config,
+    };
+    return {
+      refId: 'sparkline',
+      fields: [
+        sparkline.x ?? IndexVector.newField(length),
+        {
+          ...sparkline.y,
+          config: yFieldConfig,
+        },
+      ],
+      length,
+    };
+  };
+
+  prepareConfig = (data: DataFrame, props: Props) => {
```

Comment:
I have slight doubts about Sparkline implementing the entire config preparation. What about constructing the entire dataframe in the Sparkline, and only passing it to GraphNG component so that it can handle the config creation?

Reply:
but this is not using GraphNG (and I do not think it should)

Reply:
hm.. now I look again, see your point a bit more. There is a lot of code here to build the config that is similar to what is in GraphNG

Reply:
I agree we we should avoid duplicated code -- but in this case everything could/should be in the UPlotConfigBuilder.  I think the distinction here should be:

* GraphNG
  * DataFrame[] >>> uplot configs, uplot data, legend
* GraphNG
  * Sparkline( x, y field ) >>> uplot configs, uplot data